src/runtime/runtime_state.cs

```diff
@@ -0,0 +1,236 @@
+            if (ShouldRestoreObjects)
+            {
+                RestoreObjects(dummyGC);
+            }
+        }
+
+        private static void ResotreModules(IntPtr dummyGC)
+        {
+            var intialModules = PySys_GetObject("initial_modules");
+            Debug.Assert(intialModules != null);
+            var modules = PyImport_GetModuleDict();
+            foreach (var name in GetModuleNames())
+            {
+                if (PySet_Contains(intialModules, name) == 1)
+                {
+                    continue;
+                }
+                var module = PyDict_GetItem(modules, name);
+
+                if (UseDummyGC && _PyObject_GC_IS_TRACKED(module))
+                {
+                    ExchangeGCChain(module, dummyGC);
+                }
+                if (PyDict_DelItem(modules, name) != 0)
+                {
+                    PyErr_Print();
+                }
+            }
+        }
+
+        private static void RestoreObjects(IntPtr dummyGC)
+        {
+            if (!UseDummyGC)
+            {
+                throw new Exception("To prevent crash by _PyObject_GC_UNTRACK in Python internal, UseDummyGC should be enabled when using ResotreObjects");
+            }
+            var intialObjs = PySys_GetObject("initial_objs");
+            Debug.Assert(intialObjs != null);
+            foreach (var obj in PyGCGetObjects())
+            {
+                var p = PyLong_FromVoidPtr(obj);
+                try
+                {
+                    if (PySet_Contains(intialObjs, p) == 1)
+                    {
+                        continue;
+                    }
+                }
+                finally
+                {
+                    XDecref(p);
+                }
+                Debug.Assert(_PyObject_GC_IS_TRACKED(obj), "A GC object must be tracked");
+                ExchangeGCChain(obj, dummyGC);
+            }
+        }
+
+        public static IEnumerable<IntPtr> PyGCGetObjects()
+        {
+            var gc = PyImport_ImportModule("gc");
+            var get_objects = PyObject_GetAttrString(gc, "get_objects");
+            var objs = PyObject_CallObject(get_objects, IntPtr.Zero);
+            var length = PyList_Size(objs);
+            for (long i = 0; i < length; i++)
+            {
+                var obj = PyList_GetItem(objs, i);
+                yield return obj;
+            }
+            XDecref(objs);
+            XDecref(gc);
+        }
+
+        public static IEnumerable<IntPtr> GetModuleNames()
+        {
+            var modules = PyImport_GetModuleDict();
+            var names = PyDict_Keys(modules);
+            var length = PyList_Size(names);
+            for (int i = 0; i < length; i++)
+            {
+                var name = PyList_GetItem(names, i);
+                yield return name;
+            }
+        }
+
+        private static void AddObjPtrToSet(IntPtr set, IntPtr obj)
+        {
+            var p = PyLong_FromVoidPtr(obj);
```

Comment:
Curious, why do you have to keep these objects in the Python Set instead of .NET `HashSet`? Do they need to survive domain reload?

Reply:
Yes, they do.